### PR TITLE
Adding the Binary Paint Shop Problem

### PR DESCRIPTION
**Pull request by vijeycreative**
Added the Binary Paint Shop Problem (BPSP) problem class with its corresponding tests. This implementation was based on the following works:
1. [Beating Classical Heuristics for the Binary Paint Shop Problem with the Quantum Approximate Optimization Algorithm](https://journals.aps.org/pra/abstract/10.1103/PhysRevA.104.012403)
2. [Some Heuristics for the Binary Paint Shop Problem and their Expected Number of Colour Changes](https://www.sciencedirect.com/science/article/pii/S1570866710000559)
3. Upcoming/Unpublished work by V Vijendran et al from A*STAR and CQC2T.

## Description

This pull request introduces support for the Binary Paint Shop Problem, a combinatorial optimization task inspired by real-world automotive paint shop scenarios. In this challenge, a specific car sequence exists where each car appears twice. The main goal is to determine the most efficient paint sequence to minimize the number of paint swaps between consecutive cars.

The addition includes a new `BPSP` class within the problem module. This class facilitates the creation of Binary Paint Shop Problem instances. It offers methods to:

1.  Generate random BPSP problem instances.
2.  Convert BPSP instances into their related graph structures.
3.  Solve the problem through traditional heuristics like the Red-First and Greedy Heuristic, as well as using the CPLEX and QUBO solvers.
4.  Visualize the paint sequence for any given car sequence based on the solution provided.

Example:
```python  
# Required libraries are imported
import matplotlib.pyplot as plt
from openqaoa.problems import BPSP
from openqaoa.algorithms import QAOA

# A random instance of the Binary Paint Shop Problem is generated with 16 cars
bpsp = BPSP.random_instance(num_cars=16, seed=42)

# Solutions are computed using the CPLEX, Red-First, and Greedy algorithms
cplex_seq, cplex_swaps = bpsp.solve_cplex()
redfirst_seq, redfirst_swaps = bpsp.solve_redfirst()
greedy_seq, greedy_swaps = bpsp.solve_greedy()

# QAOA (Quantum Approximate Optimization Algorithm) is initialized and compiled
qaoa = QAOA()
qaoa.compile(bpsp.qubo)
qaoa.optimize()
# The result from QAOA is fetched, with a focus on the lowest-cost solution
qaoa_bitstring = qaoa.result.lowest_cost_bitstrings()["solutions_bitstrings"][0]
qaoa_seq, qaoa_swaps = bpsp.paintseq_from_bits(qaoa_bitstring)

# Creating a plotting area with 4 rows (one for each method)
fig, ax = plt.subplots(4, 1, figsize=(20, 8))

# Plotting the paint sequence for each of the algorithms used
bpsp.plot_paint_sequence(redfirst_seq, ax[0])
ax[0].set_title(f"Red-First Paint Sequence: {redfirst_swaps} Swaps")
bpsp.plot_paint_sequence(greedy_seq, ax[1])
ax[1].set_title(f"Greedy Paint Sequence: {greedy_swaps} Swaps")
bpsp.plot_paint_sequence(cplex_seq, ax[2])
ax[2].set_title(f"CPLEX Paint Sequence: {int(cplex_swaps)} Swaps")
bpsp.plot_paint_sequence(qaoa_seq, ax[3])
ax[3].set_title(f"QAOA Paint Sequence: {qaoa_swaps} Swaps")

# Adjusting the vertical spacing between plots to avoid overlap
fig.subplots_adjust(hspace=0.75)

# Displaying the final combined plot
plt.show()
```
image

## Checklist

[//]: <> (- [ ] My code follows the style guidelines of this project)

- [x] I have performed a self-review of my code.
- [x] I have commented my code and used numpy-style docstrings
- [ ] I have made corresponding updates to the documentation.
- [x] My changes generate no new warnings
- [x] I have added/updated tests to make sure bugfix/feature works.
- [x] New and existing unit tests pass locally with my changes.

[//]: <> (- [ ] Any dependent changes have been merged and published in downstream modules)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

The tests for this `BPSP` class are located within `./tests/test_bpsp.py` and `./tests/test_qubo.py`.

### CODE REVIEW

File: src/openqaoa-core/openqaoa/problems/bpsp.py [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+        seed = kwargs.get("seed", None)  # Default value is None if not provided.
+
+        # Assert that the number of cars is greater than 0.
+        assert num_cars > 0, "The number of cars must be greater than 0."
+
+        # Generate a list with two occurrences of each car ID, i.e., [0, 1, ..., n, 0, 1, ..., n].
+        car_sequence = np.array(list(range(num_cars)) + list(range(num_cars)))
+
+        # Set the seed for numpy's random module.
+        np.random.seed(seed)
```
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/problems/bpsp.py:203]: Use `rng = np.random.default_rng(seed)` and `rng.integers()` to create random numbers

File: src/openqaoa-core/openqaoa/problems/bpsp.py [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+        # Extract the solution values representing paint choices
+        solution = [int(np.round(model.solution.get_value(var))) for var in model.iter_binary_vars()]
+        
+        # Get the objective value of the solution
+        diff_sum = lambda lst: sum(abs(lst[i] - lst[i+1]) for i in range(len(lst)-1))
+        objective_value = diff_sum(solution)
+        # Return the paint choices and their corresponding objective value
+        return solution, objective_value
+
+    def qaoa_solution(self, bitstring):
```
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/problems/bpsp.py:438]: Change the name of the function so it reflects its action, which is converting the solution bitstring/basis statevector into a paint sequence

File: src/openqaoa-core/openqaoa/problems/bpsp.py [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+
+        # Here, we use self.bpsp_graph to gather terms and weights.
+        for u, v, edge_weight in self.bpsp_graph.edges(data="weight"):
+            terms.append([u, v])
+
+            weights.append(edge_weight)  
+
+        return QUBO(self.bpsp_graph.number_of_nodes(), terms, weights, self.problem_instance,)
+
+    def cplex_solution(self):
```
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/problems/bpsp.py:404]: Change the name of the function so it reflects its action, e.g. `find_cplex_solution()` or `solve_cplex()`

File: src/openqaoa-core/openqaoa/problems/bpsp.py [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+            
+            paint_sequence[pos1] = color
+            paint_sequence[pos2] = 1 - color  # The opposite color
+        
+        # Compute the number of paint swaps in the sequence
+        color_swaps = sum(abs(paint_sequence[i] - paint_sequence[i + 1]) for i in range(len(paint_sequence) - 1))
+        
+        return paint_sequence, color_swaps
+
+    def redfirst_solution(self):
```
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/problems/bpsp.py:477]: Change the name of the function so it reflects its action, e.g. `solve_redfirst()` or `find_solution_redfirst()`

File: src/openqaoa-core/openqaoa/problems/bpsp.py [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+                cars_painted[car] = True
+            else:
+                paint_sequence.append(0)
+
+        # Compute the number of color swaps in the sequence
+        color_swaps = sum(abs(paint_sequence[i] - paint_sequence[i + 1]) for i in range(len(self.car_sequence) - 1))
+            
+        return paint_sequence, color_swaps
+
+    def greedy_solution(self):
```
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/problems/bpsp.py:514]: Change the name of the function so it reflects its action

File: src/openqaoa-core/tests/test_bpsp.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+class TestBPSP(unittest.TestCase):
+    """
+    Test suite for the BPSP problem.
+
+    This test suite checks the functionality of methods related to the BPSP problem,
+    including QUBO creation, car sequencing, solution methods, and more.
+    """
+
+    def setUp(self):
```
Comment (KilianPoirier) [inline src/openqaoa-core/tests/test_bpsp.py:29]: Having this internal `self.bp` is slightly confusing considering we sometimes test using it and sometimes create a new problem.
This comment can be applied to the rest of the tests that use `self.bp`

**Review by KilianPoirier — CHANGES_REQUESTED** [review]
Here is my first review of your PR, please have a look at the comments and requested changes. Feel free to discuss them if you think your design choices make more sense.

File: src/openqaoa-core/tests/test_bpsp.py (hunk 6 above)
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_bpsp.py:29]: Do you want me to change the name of self variable, or not use a self variable at all?

Comment (KilianPoirier) [reply src/openqaoa-core/tests/test_bpsp.py:29]: The name is fine, I'm more concerned about the readability. The issue is that some tests are performed on `self.bp`, while others are made on a new problem class. 
I think it makes more sense to create a problem class in each test and then the relevant functionality of your problem class.

File: src/openqaoa-core/tests/test_bpsp.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    if len(terms_list1) != len(terms_list2):
+        bool = False
+    else:
+        for term1, term2 in zip(terms_list1, terms_list2):
+            bool = True if (term1 == term2 or term1 == term2[::-1]) else False
+
+    return bool
+
+
+class TestBPSP(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_bpsp.py:21]: Renamed all tests as `test_bpsp_WHAT_IS_TESTED`.

File: src/openqaoa-core/tests/test_qubo.py [hunk 8]
```diff
@@ -17,6 +17,7 @@
     BinPacking,
     SK,
     KColor,
+    BPSP
```
[the remark this answers is not in the document]
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_qubo.py]: Added Comma after BPSP.

File: src/openqaoa-core/tests/test_qubo.py [hunk 9]
```diff
@@ -71,6 +72,7 @@ def __generate_random_problems(self):
             "k_color": KColor.random_instance(
                 n_nodes=int(rng.integers(3, 8)), k=int(rng.integers(2, 5)), seed=seed
             ),
+            "binary_paint_shop_problem": BPSP.random_instance(num_cars = 10, seed = seed),
```
[the remark this answers is not in the document]
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_qubo.py]: Removed Extra Spaces.

File: src/openqaoa-core/tests/test_qubo.py [hunk 10]
```diff
@@ -157,6 +159,7 @@ def test_problem_instance(self):
             ],
             "sherrington_kirkpatrick": ["problem_type", "G"],
             "k_color": ["problem_type", "G", "k", "penalty"],
+            "binary_paint_shop_problem": ["problem_type", 'car_sequence', 'car_positions', 'bpsp_graph'],
```
[the remark this answers is not in the document]
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_qubo.py]: Fixed strings for consistency: `"binary_paint_shop_problem": ["problem_type", "car_sequence", "car_positions", "bpsp_graph"]`

File: src/openqaoa-core/openqaoa/problems/bpsp.py (hunk 1 above)
Comment (vijeycreative) [reply src/openqaoa-core/openqaoa/problems/bpsp.py]: Replaced `np.random.seed(seed)` with `rng = np.random.default_rng(seed)` and `rng.integers()` as shown below:

```
# Generate a list with two occurrences of each car ID, i.e., [0, 1, ..., n, 0, 1, ..., n].
        car_sequence = np.array(list(range(num_cars)) + list(range(num_cars)))

        # Create a new instance of the default random number generator.
        rng = np.random.default_rng(seed)

        # Apply the Fisher-Yates shuffle to the car_sequence.
        # Start from the end of the list and swap the current element with a randomly chosen earlier element.
        for i in range(len(car_sequence) - 1, 0, -1):
            # Select a random index between 0 and i (inclusive) using the rng instance.
            j = rng.integers(0, i + 1)
            # Swap the elements at indices i and j.
            car_sequence[i], car_sequence[j] = car_sequence[j], car_sequence[i]
```

If this was what you requested, please resolve this issue. Thanks :)

File: src/openqaoa-core/openqaoa/problems/bpsp.py (hunk 2 above)
Comment (vijeycreative) [reply src/openqaoa-core/openqaoa/problems/bpsp.py]: Changed the name of the function to `paintseq_from_bits`.

File: src/openqaoa-core/openqaoa/problems/bpsp.py (hunk 3 above)
Comment (vijeycreative) [reply src/openqaoa-core/openqaoa/problems/bpsp.py]: Changed the name of the function to `solve_cplex`.

File: src/openqaoa-core/openqaoa/problems/bpsp.py (hunk 4 above)
Comment (vijeycreative) [reply src/openqaoa-core/openqaoa/problems/bpsp.py]: Changed the name of the function to `solve_redfirst`.

File: src/openqaoa-core/tests/test_bpsp.py (hunk 6 above)
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_bpsp.py]: I have removed the `setUp` function and the `self.bp = BPSP(np.array([0, 1, 0, 1]))`. Wherever, I  have used `self.bp`, I am creating a new object `bpsp = BPSP(np.array([0, 1, 0, 1]))`.

File: src/openqaoa-core/tests/test_bpsp.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (KilianPoirier) [reply src/openqaoa-core/tests/test_bpsp.py:21]: Please change the names of the tests to reflect the new names of the functions

File: src/openqaoa-core/tests/test_bpsp.py [hunk 11]
```diff
@@ -53,69 +45,77 @@ def test_bpsp_terms_weights_constant(self):
...
         Validates that the generated random BPSP instance has one of the
         expected car sequences.
         """
-        instance = BPSP.random_instance(num_cars=2)
-        self.assertTrue(all(np.array([0, 1, 0, 1]) == instance.car_sequence) or 
-                        all(np.array([1, 1, 0, 0]) == instance.car_sequence) or 
-                        all(np.array([0, 0, 1, 1]) == instance.car_sequence) or 
-                        all(np.array([1, 0, 1, 0]) == instance.car_sequence))
+        bpsp = BPSP.random_instance(num_cars=2, seed=1234)
+        self.assertTrue(all(np.array([0, 1, 0, 1]) == bpsp.car_sequence) or
```
Comment (KilianPoirier) [inline src/openqaoa-core/tests/test_bpsp.py:69]: Please test for the exact problem instance generated, you want to check that `rng` is consistent, i.e. you need to directly compare the car sequence generated by this specific seed. One way to do so is to use the same function outside of the BPSP class to find what sequence you randomly generate.

You can refer to this piece of code in `test_maximum_cut.py`:
```
def test_maximumcut_random_problem(self):
        """Test MaximumCut random instance method"""

        seed = 1234
        gr = nx.generators.random_graphs.fast_gnp_random_graph(n=10, p=0.8, seed=seed)
        maxcut_manual_prob = MaximumCut(gr).qubo

        np.random.seed(1234)
        maxcut_random_prob = MaximumCut.random_instance(
            n_nodes=10, edge_probability=0.8, seed=seed
        ).qubo

        self.assertTrue(
            terms_list_equality(maxcut_manual_prob.terms, maxcut_random_prob.terms)
        )
        self.assertEqual(maxcut_manual_prob.weights, maxcut_random_prob.weights)
        self.assertEqual(maxcut_manual_prob.constant, maxcut_random_prob.constant)
```

File: src/openqaoa-core/tests/test_bpsp.py [hunk 12]
```diff
@@ -53,69 +45,77 @@ def test_bpsp_terms_weights_constant(self):
...
-        self.assertEqual(len(self.bp.bpsp_graph.nodes), 2)
+        bpsp = BPSP.random_instance(num_cars=2, seed=1234)
+        self.assertEqual(len(bpsp.bpsp_graph.nodes), 2)
 
-    def test_docplex_bpsp_model(self):
+    def test_bpsp_docplex_bpsp_model(self):
         """Test if the docplex model representation of BPSP is generated."""
-        self.assertIsNotNone(self.bp.docplex_bpsp_model)
+        bpsp = BPSP.random_instance(num_cars=2, seed=1234)
+        self.assertIsNotNone(bpsp.docplex_bpsp_model)
```
Comment (KilianPoirier) [inline src/openqaoa-core/tests/test_bpsp.py:87]: Check the attributes of your docplex model if possible

File: src/openqaoa-core/tests/test_bpsp.py [hunk 13]
```diff
@@ -53,69 +45,77 @@ def test_bpsp_terms_weights_constant(self):
...
-        self.assertEqual(self.bp.car_positions, {0: (0, 2), 1: (1, 3)})
+        bpsp = BPSP(np.array([0, 1, 0, 1]))
+        self.assertEqual(bpsp.car_positions, {0: (0, 2), 1: (1, 3)})
 
-    def test_graph(self):
+    def test_bpsp_graph(self):
         """Test the generation of a graph representation of the BPSP instance."""
-        self.assertEqual(len(self.bp.bpsp_graph.nodes), 2)
+        bpsp = BPSP.random_instance(num_cars=2, seed=1234)
+        self.assertEqual(len(bpsp.bpsp_graph.nodes), 2)
```
Comment (KilianPoirier) [inline src/openqaoa-core/tests/test_bpsp.py:82]: Check that the graph is exactly the graph you expect not just the number of nodes

File: src/openqaoa-core/tests/test_bpsp.py [hunk 14]
```diff
@@ -53,69 +45,77 @@ def test_bpsp_terms_weights_constant(self):
...
         """
         Test the solution of the BPSP problem using CPLEX.
 
         This test assumes that CPLEX is installed and functional. It checks
         the length of the solution and the objective value.
         """
-        solution, objective_value = self.bp.cplex_solution()
+        bpsp = BPSP(np.array([0, 1, 0, 1]))
+        solution, objective_value = bpsp.solve_cplex()
         self.assertEqual(len(solution), 4)
```
Comment (KilianPoirier) [inline src/openqaoa-core/tests/test_bpsp.py:98]: If deterministic please check the exact solution of the docplex model, if not you can ignore this

**Review by KilianPoirier — CHANGES_REQUESTED** [review]

File: src/openqaoa-core/tests/test_bpsp.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_bpsp.py:62]: I've changed the names of the functions as requested.

File: src/openqaoa-core/tests/test_bpsp.py (hunk 11 above)
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_bpsp.py]: I'm now checking if the car sequence generated by the BPSP random_instance is exactly the same as that generated by applying Fisher-Yates shuffle.

File: src/openqaoa-core/tests/test_bpsp.py (hunk 13 above)
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_bpsp.py]: I'm now checking the number of variables, number of constraints, name of the variables, checks if the variables are indeed binary, whether the model has an objective function, and if the objective mode is set to minimize.

File: src/openqaoa-core/tests/test_bpsp.py (hunk 14 above)
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_bpsp.py]: Changed the function to check for the exact solution returned by the Docplex model.

File: src/openqaoa-core/tests/test_bpsp.py (hunk 12 above)
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_bpsp.py]: I'm now checking the number of variables, number of constraints, name of the variables, checks if the variables are indeed binary, whether the model has an objective function, and if the objective mode is set to minimize.

File: src/openqaoa-core/tests/test_bpsp.py (hunk 13 above)
Comment (vijeycreative) [reply src/openqaoa-core/tests/test_bpsp.py]: I've changed the function to check if the two graphs are indeed identical by using the nx.is_isomorphic function. Just for further clarification, by checking the graph and bpsp.qubo are equal in the test_bpsp_qubo function, we are essentially checking the bpsp.graph as well.

**Pull request merged**